<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Finance Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/ADABUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/ADABUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


If you are running the lab in your  browser, so we will install the libraries using `piplite`


In [ ]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [ ]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

Now, let's import libraries that we are going to use

In [1]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser


In [2]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

ModuleNotFoundError: ignored

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

In our dataset we already have an index column, so let's use <code>index_col=0</code> parameter inside the <code>read_csv()</code> method to use first row from dataset as an index.

You can also assign the dataset to any variable you create.

</p>


In [3]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/ADABUSD_trades_1m.csv"

You will need to download the dataset; if you are running locally, please comment out the following


In [ ]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "ADABUSD_trades_1m.csv")
path="ADABUSD_trades_1m.csv"

In [4]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
df = pd.read_csv(path, index_col=0)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [ ]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


ts    open    high     low   close    volume  rec_count  \
0  2022-11-11 14:38:00  0.3450  0.3450  0.3438  0.3438  139808.0         44   
1  2022-11-11 14:39:00  0.3437  0.3441  0.3436  0.3440   80960.0         41   
2  2022-11-11 14:40:00  0.3439  0.3442  0.3432  0.3433  149624.0         42   
3  2022-11-11 14:41:00  0.3432  0.3448  0.3426  0.3446  165967.0         47   
4  2022-11-11 14:42:00  0.3446  0.3448  0.3438  0.3438   34815.0         27   

   avg_price  
0   0.344496  
1   0.343841  
2   0.343694  
3   0.343953  
4   0.344382

Let's capitalize names of the columns

In [15]:
current_columns = df.columns

capitalized_columns = [v.capitalize() for v in current_columns]
capitalized_columns

['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count', 'Avg_price']

And replace old names inside dataframe

In [17]:
df.columns = capitalized_columns
df.columns

Index(['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count',
       'Avg_price'],
      dtype='object')

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>file_name.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


<code>df.to_csv("finance.csv", index=False)</code>

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [18]:
df.dtypes


Ts            object
Open         float64
High         float64
Low          float64
Close        float64
Volume       float64
Rec_count      int64
Avg_price    float64
dtype: object

A series with the data type of each column is returned.


In [ ]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


<code>dataframe.describe()</code>

This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [6]:
df.describe()

open      high       low     close    volume  rec_count  avg_price
count  65303.00  65303.00  65303.00  65303.00  6.53e+04   65303.00   65303.00
mean       0.30      0.30      0.30      0.30  4.53e+04       9.28       0.30
std        0.03      0.03      0.03      0.03  9.96e+04       9.53       0.03
min        0.24      0.24      0.24      0.24  1.00e+00       1.00       0.24
25%        0.27      0.27      0.27      0.27  6.64e+03       4.00       0.27
50%        0.31      0.31      0.31      0.31  2.03e+04       6.00       0.31
75%        0.32      0.32      0.32      0.32  4.78e+04      12.00       0.32
max        0.36      0.36      0.36      0.36  4.87e+06     114.00       0.36

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute 'Rec_count' has 65303 counts, the mean value of this column is 9.28, the standard deviation is 9.53, the minimum value is 1, 25th percentile is 4, 50th percentile is 6, 75th percentile is 12, and the maximum value is 114. <br>

However, what if we would also like to check all the columns including those that are of type object, datetime or any other non numeric-type? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [7]:
# describe all the columns in "df" 
df.describe(include = "all")

ts      open      high       low     close    volume  \
count                 65303  65303.00  65303.00  65303.00  65303.00  6.53e+04   
unique                65303       NaN       NaN       NaN       NaN       NaN   
top     2022-11-11 14:38:00       NaN       NaN       NaN       NaN       NaN   
freq                      1       NaN       NaN       NaN       NaN       NaN   
mean                    NaN      0.30      0.30      0.30      0.30  4.53e+04   
std                     NaN      0.03      0.03      0.03      0.03  9.96e+04   
min                     NaN      0.24      0.24      0.24      0.24  1.00e+00   
25%                     NaN      0.27      0.27      0.27      0.27  6.64e+03   
50%                     NaN      0.31      0.31      0.31      0.31  2.03e+04   
75%                     NaN      0.32      0.32      0.32      0.32  4.78e+04   
max                     NaN      0.36      0.36      0.36      0.36  4.87e+06   

        rec_count  avg_price  
count    65303.00   65303.00  
unique        NaN        NaN  
top           NaN        NaN  
freq          NaN        NaN  
mean         9.28       0.30  
std          9.53       0.03  
min          1.00       0.24  
25%          4.00       0.27  
50%          6.00       0.31  
75%         12.00       0.32  
max        114.00       0.36

<p>
Now it provides the statistical summary of all the columns, including attributes that are not numeric.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


Note: you can change the precision for values using <code>pd.set_option("display.precision", 2)</code> to display only 2 numbers after the <code>.</code>

In [5]:
precision = 2
pd.set_option("display.precision", precision)

Now let's run <code>describe()</code> method one more time to make sure the precision was changed 

In [19]:
df.describe()

Open      High       Low     Close    Volume  Rec_count  Avg_price
count  65303.00  65303.00  65303.00  65303.00  6.53e+04   65303.00   65303.00
mean       0.30      0.30      0.30      0.30  4.53e+04       9.28       0.30
std        0.03      0.03      0.03      0.03  9.96e+04       9.53       0.03
min        0.24      0.24      0.24      0.24  1.00e+00       1.00       0.24
25%        0.27      0.27      0.27      0.27  6.64e+03       4.00       0.27
50%        0.31      0.31      0.31      0.31  2.03e+04       6.00       0.31
75%        0.32      0.32      0.32      0.32  4.78e+04      12.00       0.32
max        0.36      0.36      0.36      0.36  4.87e+06     114.00       0.36

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'Low' and 'High'.

</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Low', 'High']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [9]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65303 entries, 0 to 65302
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ts         65303 non-null  object 
 1   open       65303 non-null  float64
 2   high       65303 non-null  float64
 3   low        65303 non-null  float64
 4   close      65303 non-null  float64
 5   volume     65303 non-null  float64
 6   rec_count  65303 non-null  int64  
 7   avg_price  65303 non-null  float64
dtypes: float64(6), int64(1), object(1)
memory usage: 4.5+ MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
